NRF: Implement i2c interface

deadprogram opened the pull request:
This PR adds the I2C implementation for NRF. It should work for both NRF52 and NRF51, but I have only tested with the former, so far.

aykevl commented on src/examples/blinkm/blinkm.go:
```diff
@@ -20,20 +20,18 @@ func main() {
 
 	count := 0
 	for {
-		machine.I2C0.WriteTo(0x09, []byte("n"))
```
This change in examples/blinkm looks good, but should be done in a different PR. It appears to be unrelated to the nrf implementation.
In general, I prefer such examples in different PRs / commits, but when it is implemented for the first time (like on AVR) it's useful to have a real program to test it so that's why I was fine with it.

deadprogram replied:
Actually it is needed here. The AVR is forgiving enough that the extra STOP did not affect it, but it caused problems on the nrf. Changing this works on both.

aykevl commented on src/machine/machine_nrf.go:
```diff
@@ -96,3 +97,96 @@ func handleUART0() {
+	}
+
+	// Approx. 300 us needed after ending write before sending stop.
+	time.Sleep(300 * time.Microsecond)
```
That's a rather magical constant. I'm pretty sure there must be a better way than just waiting for some time (which may or may not be correct in all cases, and is probably not the most efficient way to do it).

Also, `time.Sleep()` is somewhat special. When using goroutines, it will switch to a different goroutine if one is runnable so it may not always do what you think it does.

I know that I2C is a bit strange on the nrf. I've played with it before and it was hard to get working correctly. There probably is a way with enough trial and error. For a working (but rather complex) driver implementation see https://github.com/NordicSemiconductor/nrfx/blob/master/drivers/src/nrfx_i2s.c.

deadprogram replied:
The code here https://github.com/aykevl/tinygo/pull/37/files#diff-194bcd9b39ab8cf699c207bc6242f403R181 is supposed to wait for the TX to complete. Which it does, but issuing a STOP immediately afterward locks up the I2C bus.

I discovered that the pause was needed there by accident due to some logging I had added. Afterward I found that the smallest pause that would work was the 300 us. No doubt that is somehow related to the bus speed, but it is not completely clear how, as of yet.

The nrfx library does not uses the direct registers, so it is of limited usefulness. Worth noting is that the TWIM interface did not work only the TWI. Not sure exactly why that might be, something to do with DMA that you would probably be better equipped to figure out. This PR uses TWI which works on both nrf52 and nrf51.

aykevl commented on src/machine/machine_nrf.go:
```diff
@@ -96,3 +97,96 @@ func handleUART0() {
+	// Assume stop after read.
+	i2c.Stop()
+
+	return
```
This return seems to be unnecessary?

deadprogram replied:
sure.

aykevl commented:
> https://travis-ci.com/aykevl/tinygo/jobs/149706752#L1088

Right. That is a less obvious import cycle. This is one of the reasons why I'd like to keep the machine package free of dependencies. Either that, or some other way needs to be thought up for using stdout in the runtime.

deadprogram commented:
OK so then how to handle the need to pause here without using `time`?

deadprogram commented:
OK this is interesting. I just removed the extra `time.Sleep()` because not seemingly needed by itself.

I can no longer reproduce whatever conditions were requiring that extra delay. Perhaps it was the extra pull-up resistors I had previously added which I removed from my breadboard.

In any case, please review the revised PR. Thanks.

deadprogram commented:
OK last change: you can also now set the SDA/SCL pins on the different buses so you can use them both at the same time, should you so require. They use the board-specific settings as a sensible default.

deadprogram commented:
Tested with a slightly modified version of colorlamp to match the different pin mappings on the pca10040 vs the Arduino, and it worked perfectly.

deadprogram commented on src/machine/machine_nrf.go:
```diff
@@ -96,3 +96,109 @@ func handleUART0() {
+// There are 2 I2C interfaces on the NRF.
+var (
+	I2C0 = I2C{Bus: 0}
+	I2C1 = I2C{Bus: 1}
```
Good idea. Done.

deadprogram commented on src/machine/i2c.go:
```diff
@@ -0,0 +1,20 @@
+type I2CConfig struct {
+	Frequency uint32
+	SCL       uint8
+	SDA       uint8
```
OK, sure. Change made.

deadprogram commented on src/machine/machine_nrf.go:
```diff
@@ -96,3 +96,109 @@ func handleUART0() {
+		config.SDA = SDA_PIN
+	}
+	if config.SCL == 0 {
+		config.SCL = SCL_PIN
```
Done.

deadprogram commented on src/examples/blinkm/blinkm.go:
```diff
@@ -20,20 +20,18 @@ func main() {
 		case 0:
 			// Crimson
-			machine.I2C0.WriteTo(0x09, []byte{0xdc, 0x14, 0x3c})
+			machine.I2C0.WriteTo(0x09, []byte{byte('n'), 0xdc, 0x14, 0x3c})
```
You are correct. Done.

deadprogram commented:
OK updated from all feedback items. Ready for review.

deadprogram commented:
Approved but not merged?

aykevl commented:
I was busy with the merge, which failed at first due to mixing ADC/PWM commits.
Merged in e4fa1a8288f7f0d539cbf0df26d554cac89ab8a6.

deadprogram commented:
:bowing_man: thanks.
